### PR TITLE
Add generic test for classify method of concrete openml providers

### PR DESCRIPTION
**Pull request by shengwangsw**
The abstract test method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue() is done, it is able be called by all the test classes that extend it, which were locally fixed.

### CODE REVIEW

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 1]
```diff
@@ -78,6 +79,29 @@
...
      * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
+
+    /**
+     * Checks the method #classify() if it ensures correctly the index of the list's maximum value
+     * obtained from the method #getClassDistrubution().
+     *
+     * @throws ModelLoadingException If anything goes wrong during loading.
+     * @throws ModelTrainingException If anything goes wrong during training.
+     *
```
Comment (MiguelSCruz) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: This line is not necessary

Comment (MiguelSCruz) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java:86]: This line is not necessary

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 2]
```diff
@@ -78,6 +79,29 @@
      *                       TESTS                       *
      * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
+
+    /**
+     * Checks the method #classify() if it ensures correctly the index of the list's maximum value
+     * obtained from the method #getClassDistrubution().
```
Comment (MiguelSCruz) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: This explanation is a bit confusing. Can you take a better look at it?

**Comment by nmldiegues** [comment]
Please fix the commit title --- seems to have lost its format somehow

**Comment by nmldiegues** [comment]
Also, the commit title does not make much sense. You want to, instead, mention something that's generic and applicable to the openml-api (Which is where you are doing this).
Something like: Add generic test for classify method of concrete openml providers

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 3]
```diff
@@ -78,6 +79,29 @@
      *                       TESTS                       *
      * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
+
+    /**
+     * Checks the method #classify() if it ensures that the providers classify
```
Comment (nmldiegues) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: "if it" -> "to ensure"

**Review by nmldiegues — COMMENTED** [review]
Just a typo; otherwise LGTM

**Comment by shengwangsw** [comment]
I think all provided improvements were done.

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 4]
```diff
@@ -25,6 +25,7 @@
 import com.feedzai.openml.provider.model.MachineLearningModelLoader;
 import com.feedzai.openml.util.algorithm.MLAlgorithmEnum;
 import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Doubles;
```
Comment (nmldiegues) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: This import is unused, right? Please clean it up

**Review by nmldiegues — APPROVED** [review]

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java (hunk 4 above)
Comment (shengwangsw) [reply openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: This is used in the method AbstractProviderModelBaseTest#classifyIndexOfMaxScoresValue() to obtain the index of maximum value from a list of double, e.g. _Doubles.asList(scores).indexOf(maxScore)_.

But I just reviewed the code, and there is a method that uses the ArrayUtils. For this reason I will remove this import and update the method mentioned above to use the following _Arrays.asList(ArrayUtils.toObject(scores)).indexOf(maxScore)_.

**Comment by nmldiegues** [comment]
@Shenguito  what's preventing this from being merged + back-ported + released?
@ptraca  ?

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 5]
```diff
@@ -78,6 +78,29 @@
      *                       TESTS                       *
      * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
+
+    /**
+     * Checks the method #classify() to ensure that the providers classify
+     * correctly the index of the maximum value of the scores list.
+     *
+     * @throws ModelLoadingException  If anything goes wrong during loading.
```
Comment (ptraca) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: I don't think we need to document the @Throws clause in test methods.

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 6]
```diff
@@ -78,6 +78,29 @@
      *                       TESTS                       *
      * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
+
+    /**
+     * Checks the method #classify() to ensure that the providers classify
```
Comment (ptraca) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: It should be something like

```java
    /**
     * Verifies that the {@link ClassificationMLModel#classify(Instance)} " returns the index of the greatest value in 
     * the class probability distribution produced by the calling 
     * {@link ClassificationMLModel#getClassDistribution(Instance)} on the model
     * 
     * @see ClassificationMLModel
     */

```

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 7]
```diff
@@ -78,6 +78,29 @@
...
+     * correctly the index of the maximum value of the scores list.
+     *
+     * @throws ModelLoadingException  If anything goes wrong during loading.
+     * @throws ModelTrainingException If anything goes wrong during training.
+     */
+    @Test
+    public void classifyIndexOfMaxScoresValue() throws ModelLoadingException, ModelTrainingException {
```
Comment (ptraca) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: I think it should be called `canGetClassDistributionMaxValueIndex()`
Also we don't need the specific types of Exceptions, the super type will do.


```java
    @Test
    public void classifyIndexOfMaxScoresValue() throws Exception {
```

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java [hunk 8]
```diff
@@ -78,6 +78,29 @@
...
+
+        final M model = getFirstModel();
```
Comment (ptraca) [inline openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: if you space it like this reads a lot better.

```java
 ...
        final M model = getFirstModel();
        final Instance instance = getDummyInstance();
        final double[] scores = model.getClassDistribution(instance);

        final int classificationIndex = model.classify(instance);
        
        final double maxScore = Arrays.stream(scores).max().getAsDouble();

        assertThat(Arrays.asList(ArrayUtils.toObject(scores)).indexOf(maxScore))
                .as("The index of maximum value")
                .isEqualTo(classificationIndex);

```

**Comment by ptraca** [comment]
After reviewing the commits in `feedzai-openml` and `feedzai-openml-java`, It seems to me that there isn't any good way to test the H2O model provider implementation on `ClassificationMLModel#classify(Instance)` method. Given that the method isn't used by the platform we should mark it as @deprecated.

The unit test in `AbstractProviderModelBaseTest` can remain for now to validate those implementations that can be currently tested but @override the H20 implementation in the `feedzai-openml-java` repository and provide a dummy implementation.

Don't forget to explain in the comment the way of the empty test implementation.

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java (hunk 5 above)
Comment (shengwangsw) [reply openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: I shared the same opinion, I just followed the pattern of the other methods.
I will change it then.

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java (hunk 6 above)
Comment (shengwangsw) [reply openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: Right. I will update it.

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java (hunk 8 above)
Comment (shengwangsw) [reply openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: Indeed it seems better.

File: openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java (hunk 7 above)
Comment (shengwangsw) [reply openml-utils/src/test/java/com/feedzai/openml/util/provider/AbstractProviderModelBaseTest.java]: Ok, I will change it.

**Comment by ptraca** [comment]
You forgot to deprecate the `classify` method in the `ClassificationMLModel` interface.

@see `https://docs.oracle.com/javase/8/docs/technotes/guides/javadoc/deprecation/deprecation.html`for examples.

**Comment by shengwangsw** [comment]
My bad, I forgot to push.

**Comment by ptraca** [comment]
LGTM. good to merge.

**Comment by ptraca** [comment]
Don't forget to 

- after releasing a version o the lib to back-port it to the active branchs
- Add the back-port versions and commit SHA's to the merge to main commit
- Update the ticket status and affected version fields

This applies to this library and and the dependents (java/python/r, openml internal, etc)

**Pull request merged**